memory leaks in NAEFS code

From: Bo Cui - NOAA Affiliate <[bo.cui@noaa.gov](mailto:bo.cui@noaa.gov)>
Date: Fri, May 26, 2023 at 3:31 PM
Subject: memory leaks in NAEFS code
To: Arun Chawla - NOAA Federal <[arun.chawla@noaa.gov](mailto:arun.chawla@noaa.gov)>, Rahul Mahajan - NOAA Federal <[rahul.mahajan@noaa.gov](mailto:rahul.mahajan@noaa.gov)>
Cc: Hui-Ya Chuang - NOAA Federal <[Hui-Ya.Chuang@noaa.gov](mailto:Hui-Ya.Chuang@noaa.gov)>, Jason Levit - NOAA Federal <[jason.levit@noaa.gov](mailto:jason.levit@noaa.gov)>


Hi all,

As you know, I am debugging memory leaks in NAEFS code in Bugzilla ticket [#766](http://www2.spa.ncep.noaa.gov/bugzilla/show_bug.cgi?id=766). Here are my updates.

1. Used "-check all" to check code gefs_debias.f90 and found one array out of bound in the code, which was already fixed.

2. Continued using valgrind to check the memory leak issue.

I had a code to call getgb2 only and print out some messages. The code work fine if the default jpdt, jids and etc. were initialized. Once  jpdt and jpdtn were given values ( the same process to unpack grib2 message as the NAEFS code), the valgrind output will show memory leak problems. The memory leak comes from subroutine getgb2.
 
==249219== 18,952 bytes in 206 blocks are definitely lost in loss record 1 of 2
==249219==    at 0x4C33221: malloc (vg_replace_malloc.c:307)
==249219==    by 0x544F9F: _mm_malloc (in /lfs/h2/emc/vpppg/noscrub/bo.cui/test_getgb2/test_getgb2)
==249219==    by 0x4E029D: for_allocate (in /lfs/h2/emc/vpppg/noscrub/bo.cui/test_getgb2/test_getgb2)
==249219==    by 0x411ABB: gf_unpack1_ (gf_unpack1.f:79)
==249219==    by 0x4105DC: getgb2s_ (getgb2s.f:362)
==249219==    by 0x40F0BD: getgb2_ (getgb2.f:318)
==249219==    by 0x404164: gb2read_ (test_getgb2.f:215)
==249219==    by 0x40AE6F: MAIN__ (test_getgb2.f:67)
==249219==    by 0x403251: main (in /lfs/h2/emc/vpppg/noscrub/bo.cui/test_getgb2/test_getgb2)

More details can be found on cactus:
/lfs/h2/emc/vpppg/noscrub/bo.cui/test_getgb2/valgrind-out_test_getgb2.txt
 
Script to run the executable:
/lfs/h2/emc/vpppg/noscrub/bo.cui/test_getgb2/job_test_getgb2

sorc code:
/lfs/h2/emc/vpppg/noscrub/bo.cui/test_getgb2/test_getgb2.fd
G2 Lib is from:
/apps/ops/prod/libs/build/v1.1.0/pkg/g2-v3.4.1/src

I check wafs_belending codes and some other operational codes. So far I haven't found any special steps required before calling getgb2. Any suggestions and comments for this issue?

@edwardhartnett @Hang-Lei-NOAA  I added a line "call  gf_finalize()" at the end of my code and re-compile it. The compilation failed with the error message:

 undefined reference to `gf_finalize_'

I tested code using new g2lib and prod g2lib, both failed with the same error. I checked codes on wcoss2 at /apps/ops/prod/libs/build/v1.1.0/pkg/g2-v3.4.1/src

There is no subroutine related to gf_finalize. Please correct me if I am wrong or missing something. Thanks!

Well sure, because I haven't merged the PR yet, and Hang has not yet
installed a new version. ;-)

When that happens we will let you know. I'll get the PR merged Monday
morning, and then Hang can install it for you...

Thanks,
Ed

On Sun, Jul 16, 2023 at 12:08 PM BoCui-NOAA ***@***.***>
wrote:

> @edwardhartnett <https://github.com/edwardhartnett> @Hang-Lei-NOAA
> <https://github.com/Hang-Lei-NOAA> I added a line "call gf_finalize()" at
> the end of my code and re-compile it. The compilation failed with the error
> message:
>
> undefined reference to `gf_finalize_'
>
> I tested code using new g2lib and prod g2lib, both failed with the same
> error. I checked codes on wcoss2 at
> /apps/ops/prod/libs/build/v1.1.0/pkg/g2-v3.4.1/src
>
> There is no subroutine related to gf_finalize. Please correct me if I am
> wrong or missing something. Thanks!
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/NOAA-EMC/NCEPLIBS-g2/issues/446#issuecomment-1637154117>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AJIOMMAODEPEI7XPMWNC2ILXQQU2NANCNFSM6AAAAAAZJEAPOM>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>


-- 
Edward Hartnett
CIRES/NOAA EMC

> Well sure, because I haven't merged the PR yet, and Hang has not yet installed a new version. ;-) When that happens we will let you know. I'll get the PR merged Monday morning, and then Hang can install it for you... Thanks, Ed
> […](#)
> On Sun, Jul 16, 2023 at 12:08 PM BoCui-NOAA ***@***.***> wrote: @edwardhartnett <https://github.com/edwardhartnett> @Hang-Lei-NOAA <https://github.com/Hang-Lei-NOAA> I added a line "call gf_finalize()" at the end of my code and re-compile it. The compilation failed with the error message: undefined reference to `gf_finalize_' I tested code using new g2lib and prod g2lib, both failed with the same error. I checked codes on wcoss2 at /apps/ops/prod/libs/build/v1.1.0/pkg/g2-v3.4.1/src There is no subroutine related to gf_finalize. Please correct me if I am wrong or missing something. Thanks! — Reply to this email directly, view it on GitHub <[#446 (comment)](https://github.com/NOAA-EMC/NCEPLIBS-g2/issues/446#issuecomment-1637154117)>, or unsubscribe <https://github.com/notifications/unsubscribe-auth/AJIOMMAODEPEI7XPMWNC2ILXQQU2NANCNFSM6AAAAAAZJEAPOM> . You are receiving this because you were mentioned.Message ID: ***@***.***>
> -- Edward Hartnett CIRES/NOAA EMC

Got it, will wait for Hang's confirmation. Thanks!

--- a/src/getidx.F90
+++ b/src/getidx.F90
@@ -40,12 +40,11 @@ subroutine getidx(lugb, lugi, cindex, nlen, nnum, iret)
   integer, intent(in) :: lugb, lugi
   integer, intent(out) :: nlen, nnum, iret
   character(len = 1), pointer, dimension(:) :: cindex
-
   integer, parameter :: maxidx = 10000
   integer, parameter :: msk1 = 32000, msk2 = 4000
 
   integer :: lux
-  integer :: irgi, mskp, nmess
+  integer :: irgi, mskp, nmess, i
 
   type gindex
      integer :: nlen
@@ -72,6 +71,20 @@ subroutine getidx(lugb, lugi, cindex, nlen, nnum, iret)
      end subroutine getg2ir
   end interface
 
+  ! Free all associated memory and exit.
+  if (lugb .eq. 0 .and. lugi .eq. 0) then
+     print *, 'getidx: Freeing all memory'
+     do i = 1, 10000
+        if (associated(idxlist(i)%cbuf)) then
+           print *, 'deallocating ', loc(idxlist(i)%cbuf)
+           deallocate(idxlist(i)%cbuf)
+           nullify(idxlist(i)%cbuf)
+        endif
+     end do
+     iret = 0
+     return
+  endif
+
   !  determine whether index buffer needs to be initialized
   lux = 0
   iret = 0
@@ -84,6 +97,7 @@ subroutine getidx(lugb, lugi, cindex, nlen, nnum, iret)
   if (lugi .eq. lugb) then      ! force regeneration of index from grib2 file
      if (associated(idxlist(lugb)%cbuf))  &
           deallocate(idxlist(lugb)%cbuf)
+     print *, 'Force regeneration'
      nullify(idxlist(lugb)%cbuf)
      idxlist(lugb)%nlen = 0
      idxlist(lugb)%nnum = 0
@@ -94,6 +108,7 @@ subroutine getidx(lugb, lugi, cindex, nlen, nnum, iret)
      ! associated with unit abs(lugi)
      if (associated(idxlist(lugb)%cbuf))  &
           deallocate(idxlist(lugb)%cbuf)
+     print *, 'Force re-read'
      nullify(idxlist(lugb)%cbuf)
      idxlist(lugb)%nlen = 0
      idxlist(lugb)%nnum = 0
@@ -102,6 +117,7 @@ subroutine getidx(lugb, lugi, cindex, nlen, nnum, iret)
 
   !  check if index already exists in memory
   if (associated(idxlist(lugb)%cbuf)) then
+     print *, 'Index exists in memory!'
      cindex => idxlist(lugb)%cbuf
      nlen = idxlist(lugb)%nlen
      nnum = idxlist(lugb)%nnum
@@ -128,3 +144,24 @@ subroutine getidx(lugb, lugi, cindex, nlen, nnum, iret)
      return
   endif
 end subroutine getidx
+
+!> Free all memory associated with the library.
+!>
+!> @author Ed Hartnett @date 7/16/23
+subroutine gf_finalize()
+  implicit none
+
+  character(len = 1), pointer, dimension(:) :: cindex
+  integer :: nlen, nnum, iret
+  ! Declare interfaces (required for cbuf pointer).
+  interface
+     subroutine getidx(lugb,lugi,cbuf,nlen,nnum,irgi)
+       character(len=1),pointer,dimension(:) :: cbuf
+       integer,intent(in) :: lugb,lugi
+       integer,intent(out) :: nlen,nnum,irgi
+     end subroutine getidx
+  end interface
+
+  call getidx(0, 0, cindex, nlen, nnum, iret)
+
+end subroutine gf_finalize

--- a/src/getidx.F90
+++ b/src/getidx.F90
@@ -72,11 +72,11 @@ subroutine getidx(lugb, lugi, cindex, nlen, nnum, iret)
   end interface
 
   ! Free all associated memory and exit.
-  if (lugb .eq. 0 .and. lugi .eq. 0) then
-     print *, 'getidx: Freeing all memory'
+  if (lugb .eq. 0) then
+     !print *, 'getidx: Freeing all memory'
      do i = 1, 10000
         if (associated(idxlist(i)%cbuf)) then
-           print *, 'deallocating ', loc(idxlist(i)%cbuf)
+           !print *, 'deallocating ', loc(idxlist(i)%cbuf)
            deallocate(idxlist(i)%cbuf)
            nullify(idxlist(i)%cbuf)
         endif
@@ -97,7 +97,7 @@ subroutine getidx(lugb, lugi, cindex, nlen, nnum, iret)
   if (lugi .eq. lugb) then      ! force regeneration of index from grib2 file
      if (associated(idxlist(lugb)%cbuf))  &
           deallocate(idxlist(lugb)%cbuf)
-     print *, 'Force regeneration'
+     !print *, 'Force regeneration'
      nullify(idxlist(lugb)%cbuf)
      idxlist(lugb)%nlen = 0
      idxlist(lugb)%nnum = 0
@@ -108,7 +108,7 @@ subroutine getidx(lugb, lugi, cindex, nlen, nnum, iret)
      ! associated with unit abs(lugi)
      if (associated(idxlist(lugb)%cbuf))  &
           deallocate(idxlist(lugb)%cbuf)
-     print *, 'Force re-read'
+     !print *, 'Force re-read'
      nullify(idxlist(lugb)%cbuf)
      idxlist(lugb)%nlen = 0
      idxlist(lugb)%nnum = 0
@@ -117,7 +117,7 @@ subroutine getidx(lugb, lugi, cindex, nlen, nnum, iret)
 
   !  check if index already exists in memory
   if (associated(idxlist(lugb)%cbuf)) then
-     print *, 'Index exists in memory!'
+     !print *, 'Index exists in memory!'
      cindex => idxlist(lugb)%cbuf
      nlen = idxlist(lugb)%nlen
      nnum = idxlist(lugb)%nnum
@@ -147,12 +147,17 @@ end subroutine getidx
 
 !> Free all memory associated with the library.
 !>
+!> @param[out] iret integer Return code:
+!> - 0 No error.
+!> - otherwise Error freeing internal resources.
 !> @author Ed Hartnett @date 7/16/23
-subroutine gf_finalize()
+subroutine gf_finalize(iret)
   implicit none
 
+  integer, intent(out) :: iret  
   character(len = 1), pointer, dimension(:) :: cindex
-  integer :: nlen, nnum, iret
+  integer :: nlen, nnum
+
   ! Declare interfaces (required for cbuf pointer).
   interface
      subroutine getidx(lugb,lugi,cbuf,nlen,nnum,irgi)
@@ -162,6 +167,8 @@ subroutine gf_finalize()
      end subroutine getidx
   end interface
 
+  ! Call getidx with 0 for the first parameter, ensuring that the
+  ! internal memory is freed.
   call getidx(0, 0, cindex, nlen, nnum, iret)
 
 end subroutine gf_finalize

--- a/tests/test_getgb2_mem.F90
+++ b/tests/test_getgb2_mem.F90
@@ -47,6 +47,9 @@ program test_getgb2_mem
      if (ios .ne. 0) stop 11
   end do
 
+  ! Free all memory.
+  call gf_finalize()
+
 end program test_getgb2_mem
 
 ! Read GRIB2 file.

--- a/tests/test_getgb2_mem.F90
+++ b/tests/test_getgb2_mem.F90
@@ -27,7 +27,7 @@ program test_getgb2_mem
   integer lugb
   parameter(lugb = 10)
   integer :: ios
-  integer :: i
+  integer :: i, iret
   
   test_file = [character(len=120) :: TEST_FILE_GEP19_BCF144, TEST_FILE_GEAVG, TEST_FILE_GEC00, &
        TEST_FILE_GEGFS_F144, TEST_FILE_GEGFS_MEF144, TEST_FILE_GEP19_F144]
@@ -48,7 +48,8 @@ program test_getgb2_mem
   end do
 
   ! Free all memory.
-  call gf_finalize()
+  call gf_finalize(iret)
+  if (iret .ne. 0) stop 12
 
 end program test_getgb2_mem
 

--- a/tests/test_getidx.F90
+++ b/tests/test_getidx.F90
@@ -41,8 +41,8 @@ program test_getidx
   if (iret .ne. 0) stop 100
 
   ! This will not work because the first argument must be between 0 and 9999.
-  call getidx(0, lugi, cbuf, nlen, nnum, iret)
-  if (iret .ne. 90) stop 99
+!  call getidx(0, lugi, cbuf, nlen, nnum, iret)
+!  if (iret .ne. 90) stop 99
   call getidx(10000, lugi, cbuf, nlen, nnum, iret)
   if (iret .ne. 90) stop 99
 

